waddrmgr/manager_test.go

```diff
@@ -2761,3 +2762,104 @@ func testNewRawAccount(t *testing.T, _ *Manager, db walletdb.DB,
 			accountTargetAddr.AddrHash())
 	}
 }
+
+// TestDeriveFromKeyPathCache tests that the DeriveFromKeyPathCache method will
+// properly cache items in the cache, and return corresponding errors if the
+// account isn't properly cached.
+func TestDeriveFromKeyPathCache(t *testing.T) {
+	t.Parallel()
+
+	teardown, db := emptyDB(t)
+	defer teardown()
+
+	// We'll start the test by creating a new root manager that will be
+	// used for the duration of the test.
+	var mgr *Manager
+	err := walletdb.Update(db, func(tx walletdb.ReadWriteTx) error {
+		ns, err := tx.CreateTopLevelBucket(waddrmgrNamespaceKey)
+		if err != nil {
+			return err
+		}
+		err = Create(
+			ns, seed, pubPassphrase, privPassphrase,
+			&chaincfg.MainNetParams, fastScrypt, time.Time{},
+		)
+		if err != nil {
+			return err
+		}
+		mgr, err = Open(ns, pubPassphrase, &chaincfg.MainNetParams)
+
+		return mgr.Unlock(ns, privPassphrase)
+	})
+	if err != nil {
+		t.Fatalf("create/open: unexpected error: %v", err)
+	}
+	defer mgr.Close()
+
+	// Now that we have the manager created, we'll fetch one of the default
+	// scopes for usage within this test.
+	scopedMgr, err := mgr.FetchScopedKeyManager(KeyScopeBIP0044)
+	if err != nil {
+		t.Fatalf("unable to fetch scope %v: %v", KeyScopeBIP0044, err)
+	}
+
+	keyPath := DerivationPath{
+		InternalAccount: 0,
+		Account:         hdkeychain.HardenedKeyStart,
+		Branch:          10,
+		Index:           1,
+	}
+
+	// Our test starts here, we'll attempt to derive a new key using the
+	// cached method. This should fail at first since the account itself
+	// isn't cached.
+	_, err = scopedMgr.DeriveFromKeyPathCache(keyPath)
+	if !IsError(err, ErrAccountNotCached) {
+		t.Fatalf("didn't get account not cached error: %v", err)
+	}
+
+	// Now we'll attempt to derive the key using the normal method that
+	// requires a database transaction.
+	var derivedKey *btcec.PrivateKey
+	err = walletdb.Update(db, func(tx walletdb.ReadWriteTx) error {
+		ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+
+		managedAddr, err := scopedMgr.DeriveFromKeyPath(
+			ns, keyPath,
+		)
+		if err != nil {
+			return err
+		}
+
+		derivedKey, err = managedAddr.(ManagedPubKeyAddress).PrivKey()
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
```

Comment:
It looks like we have `testify/require` in the dependencies already, could use it here as well.